### PR TITLE
[RLlib]: Fix `max_concurrent` in `external_env.py`

### LINKED ISSUE #31965: RLLib: ExternalEnv concurrent episode leaking issue
### What happened + What you expected to happen

I'm RLLib's external env with PPO to drive a physical system. For this reason, I am limited to one env with one worker. To ensure this, I build the algorithm (PPO) with `num_rollout_workers=1, batch_mode="complete_episodes", num_envs_per_worker=1` 

**The Bug**
And after a couple of episodes, I consistently get `AssertionError: Too many concurrent episodes, were some leaked? This ExternalEnv was created with max_concurrent=2`

The full stack trace:
```
Traceback (most recent call last):
  File "/home/aydin/Documents/TREC-RLT/scripts/pendulum/pendulum_external_replication.py", line 70, in <module>
    main()
  File "/home/aydin/Documents/TREC-RLT/scripts/pendulum/pendulum_external_replication.py", line 67, in main
    agent.train()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/tune/trainable/trainable.py", line 367, in train
    raise skipped from exception_cause(skipped)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/tune/trainable/trainable.py", line 364, in train
    result = self.step()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 749, in step
    results, train_iter_ctx = self._run_one_training_iteration()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 2623, in _run_one_training_iteration
    results = self.training_step()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/ppo/ppo.py", line 318, in training_step
    train_batch = synchronous_parallel_sample(
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/execution/rollout_ops.py", line 85, in synchronous_parallel_sample
    sample_batches = worker_set.foreach_worker(
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/worker_set.py", line 696, in foreach_worker
    handle_remote_call_result_errors(remote_results, self._ignore_worker_failures)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/worker_set.py", line 73, in handle_remote_call_result_errors
    raise r.get()
ray.exceptions.RayTaskError(AssertionError): ray::RolloutWorker.apply() (pid=195731, ip=192.168.2.49, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x7f0288f98970>)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/utils/actor_manager.py", line 183, in apply
    raise e
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/utils/actor_manager.py", line 174, in apply
    return func(self, *args, **kwargs)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/execution/rollout_ops.py", line 86, in <lambda>
    lambda w: w.sample(), local_worker=False, healthy_only=True
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/rollout_worker.py", line 900, in sample
    batches = [self.input_reader.next()]
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 92, in next
    batches = [self.get_data()]
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 285, in get_data
    item = next(self._env_runner)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 660, in _env_runner
    unfiltered_obs, rewards, dones, infos, off_policy_actions = base_env.poll()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/env/external_env.py", line 372, in poll
    assert len(results[0]) <= limit, (
AssertionError: Too many concurrent episodes, were some leaked? This ExternalEnv was created with max_concurrent=1
```

(Note than max_concurrent is set to 2, because RLLib's polling algorithm checks whether the number of concurrent episodes is `<` limit, rather than `<=` limit.

### Versions / Dependencies

OS = Ubuntu 20.04
Python = 3.9
Ray = 2.2.0
Full output of `conda list` is [here](https://github.com/ray-project/ray/files/10510898/versions.txt)

### Reproduction script

Here is the env class and the driver script. Place the env in a file `pendulum_env_replication`, and run the driver script from the same directory. 

### Driver Script
```

from ray.tune import register_env
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.env.external_env import ExternalEnv
from pendulum_env_replication import PendulumExternalEnv


def main():
        
    register_env("pendulum-v0", lambda c: PendulumExternalEnv(c))
    agent = (
        PPOConfig()
        .resources(num_gpus=1)
        .rollouts(num_rollout_workers=1, batch_mode="complete_episodes", num_envs_per_worker=1)
        .environment(env="pendulum-v0")
        .build()
    )       

    while True:
        agent.train()

if __name__ == "__main__":
    main()
```
### Env
```
from ray.rllib.env.external_env import ExternalEnv
import time
import numpy as np
import gym

class PendulumExternalEnv(ExternalEnv):
    
    
    def __init__(self, config):
        
        # action space [pwm]
        current_limit: float = 1
        self.action_space = gym.spaces.Box(low=-current_limit, high=current_limit, shape=(1,), dtype=np.float32)
        
        # observation space [current angle (rad), target angle (rad), current angular velocity (rad/s)]
        self.angle_limit: float = 0.2
        observation_limit = np.array([self.angle_limit, self.angle_limit, 2 * np.pi])
        self.observation_space = gym.spaces.Box(low=-observation_limit, high=observation_limit, dtype=np.float32)
        self.observation = np.zeros(3, dtype=np.float32)
        
        ExternalEnv.__init__(self, self.action_space, self.observation_space, max_concurrent = 1)
                
        # environment variables
        self.episode_duration: float = 5
        self.episode_num: int = 0
        self.start_time = time.time()

        
    def run(self):
        while True:
            episode_id: str = self.start_episode()
            
            while not self._check_done():
                print("loopin")
                
                action = self.get_action(episode_id, self.observation)[0]
                self.log_returns(episode_id, 0)
            
            print("resetting")
            self.start_time = time.time()
            self.end_episode(episode_id, [0,0,0]) #FIXME: add actual obs to end_episode 
            
            
    def _check_done(self):
        if time.time() - self.start_time > self.episode_duration:
            return True
        
        if abs(self.observation[1]) > self.angle_limit:
            return True
        
        return False

```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by Rohan138**
## Related issue number

Closes #31965 

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/external_env.py
Comment (avnishn) [inline rllib/env/external_env.py:381]: does this actually solve the problem? Can we add a test from the repro script for this.

Comment (Rohan138) [reply rllib/env/external_env.py:381]: This is what I'm not sure about: if this actually solves the problem or not. I was working on adding a unit test, but the test is still broken if I play with the value of `rollout_fragment_length`, so I'm confused about the original use case for the `max_concurrent` arg and when it should fail vs when it shouldn't.

Comment (Rohan138) [reply rllib/env/external_env.py:381]: Putting up a minimal repro example rn

Comment (Rohan138) [reply rllib/env/external_env.py]: Added a unit test: `python -m pytest -rP rllib/env/tests/test_external_env.py -k 'max_concurrent'` fails, basically the `rollout_fragment_length=40` setting causes the env to rollout multiple episodes (25 timesteps each) before it returns control to the algo, which conflicts with the `max_concurrent=1` setting.

My question is basically why do we even want the `max_concurrent` setting in the first place in this way. The way it's currently written, it'll also error out if we use vectorized external envs (again, multiple episodes running concurrently).

**Comment by Rohan138** [comment]
@kouroshHakha I think we should just deprecate/remove the `max_concurrent` key, which is what I've done here; it doesn't really make sense with how episodes are collected from environments (See the above discussion, I wrote a test that shows how it breaks for different `rollout_fragment_length` settings). Happy to discuss though.

**Review by kouroshHakha — COMMENTED** [review]
I don't understand what this PR is trying to achieve by removing max_concurrent parameter from the code. How is removing the code fixing issue? Is the linked issue a bug or misusage?

**Comment by Rohan138** [comment]
(fyi, one of the earlier merge commits seems to have accidentally requested review from every single code owner; please ignore)

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**
